ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/BonsaiReferenceTestWorldState.java

```diff
@@ -89,6 +92,108 @@ public ReferenceTestWorldState copy() {
+      validateFrozenStateRolling(parentStateRoot, originalUpdater, blockHeader);
+    }
+  }
+
+  private void validateFrozenStateRolling(
+      final Hash parentStateRoot,
+      final BonsaiReferenceTestUpdateAccumulator originalUpdater,
+      final BlockHeader blockHeader) {
+    // generate trie from frozen state
+    generateTrieLogFromState(blockHeader, originalUpdater, true);
+    final TrieLog trieLogFromFrozenState =
+        trieLogManager
+            .getTrieLogLayer(blockHeader.getBlockHash())
+            .orElseThrow(() -> new RuntimeException("trielog not found during test"));
+    // trying rollback rollfoward with frozen state
+    validateTrieLog(parentStateRoot, blockHeader, trieLogFromFrozenState);
+  }
+
+  private void validatePersistedStateRolling(
+      final Hash parentStateRoot,
+      final BonsaiReferenceTestUpdateAccumulator originalUpdater,
+      final BlockHeader blockHeader) {
+    // generate trie from persisted state
+    generateTrieLogFromState(blockHeader, originalUpdater, false);
+    final TrieLog trieLogFromPersistedState =
+        trieLogManager
+            .getTrieLogLayer(blockHeader.getBlockHash())
+            .orElseThrow(() -> new RuntimeException("trielog not found during test"));
+    // trying rollback rollfoward with persisted state
+    validateTrieLog(parentStateRoot, blockHeader, trieLogFromPersistedState);
+  }
+
+  private void validateTrieLog(
+      final Hash parentStateRoot, final BlockHeader blockHeader, final TrieLog trieLog) {
+
+    try (var bonsaiWorldState = createBonsaiWorldState(false)) {
+      BonsaiWorldStateUpdateAccumulator updaterForState =
+          (BonsaiWorldStateUpdateAccumulator) bonsaiWorldState.updater();
+      updaterForState.rollForward(trieLog);
+      updaterForState.commit();
+      bonsaiWorldState.persist(blockHeader);
+      Hash generatedRootHash = bonsaiWorldState.rootHash();
+      if (!bonsaiWorldState.rootHash().equals(blockHeader.getStateRoot())) {
+        throw new RuntimeException(
+            "state root becomes invalid following a rollForward %s != %s"
+                .formatted(blockHeader.getStateRoot(), generatedRootHash));
+      }
+
+      updaterForState = (BonsaiWorldStateUpdateAccumulator) bonsaiWorldState.updater();
+      updaterForState.rollBack(trieLog);
+      updaterForState.commit();
+      bonsaiWorldState.persist(null);
+      generatedRootHash = bonsaiWorldState.rootHash();
+      if (!bonsaiWorldState.rootHash().equals(parentStateRoot)) {
+        throw new RuntimeException(
+            "state root becomes invalid following a rollBackward %s != %s"
+                .formatted(parentStateRoot, generatedRootHash));
+      }
+    }
+  }
+
+  private void generateTrieLogFromState(
+      final BlockHeader blockHeader,
+      final BonsaiReferenceTestUpdateAccumulator originalUpdater,
+      final boolean isFrozen) {
+    // generate trielog
```

Comment:
perhaps add to this comment that generating a trielog is a side effect of persisting the worldstate